src/hotspot/share/runtime/synchronizer.cpp

```diff
@@ -1264,36 +1264,45 @@ static bool monitors_used_above_threshold(MonitorList* list) {
   if (MonitorUsedDeflationThreshold == 0) {  // disabled case is easy
     return false;
   }
-  // Start with ceiling based on a per-thread estimate:
-  size_t ceiling = ObjectSynchronizer::in_use_list_ceiling();
-  size_t old_ceiling = ceiling;
-  if (ceiling < list->max()) {
-    // The max used by the system has exceeded the ceiling so use that:
-    ceiling = list->max();
-  }
   size_t monitors_used = list->count();
   if (monitors_used == 0) {  // empty list is easy
     return false;
   }
-  if (NoAsyncDeflationProgressMax != 0 &&
-      _no_progress_cnt >= NoAsyncDeflationProgressMax) {
-    double remainder = (100.0 - MonitorUsedDeflationThreshold) / 100.0;
-    size_t new_ceiling = ceiling + (size_t)((double)ceiling * remainder) + 1;
```

Comment:
if you are looking fort the minimal fix for the division-by-zero problem, then I think simply fixing this line to avoid overflow will suffice:
```
size_t delta = (size_t)(ceiling * remainder) + 1;
if (ceiling > SIZE_MAX - delta) { // overflow
  ceiling = SIZE_MAX;  // Or some other positive limit
} else {
  ceiling += delta;
}
```

Reply:
I think this code with the introduction of delta looks a bit more clear to me, if you use it in your existing patch to replace lines 1286 to 1289.  That would still be a minimal fix.

Reply:
It is the only thing needed to fix the overflow. Anything else should be deferred to a later "fixing the math" issue IMO.

Reply:
What's more clear to some people, might be more unclear to some other people (like me).

To me it's clear that we have an overflow after adding two unsigned numbers A and B, if the result is less than any of the numbers A and B. A quick search on the interwebs also confirms this. So, in this case if `new_ceiling < old_ceiling` we have an overflow.

Looking at the suggested:

```
size_t delta = (size_t)(ceiling * remainder) + 1;
if (ceiling > SIZE_MAX - delta) { // overflow
```

I see "if ceiling is larger than the largest possible number minus some delta (which might be very small), we have and overflow" and I think, this might be true... Not saying you are wrong, but it's definitely not clear to me by just looking at it.

Definitely agree that we should fix the math that calculates the `new_ceiling` later.

Reply:
Both expressions: the one in the patch and the suggested expression for overflow test require study and neither are faster to understand. Both require comment so it's really a preference of which math expression you like the best. Either is fine to me because they both look correct to me and with a comment I don't have to reexamine them.

As for the minimal fix, the suggested patch does two things that are related and both desirable to have as a fix.  Moving the ceiling calculation to only be adjusted if the monitors used are above MonitorDeflationThreshold, and fixing the overflow resolve both causes of the bug.   I think we want both of these. The less minimal patch of determining how to adjust the ceiling and what to adjust it to, with what values and potentially removing some of the command line arguments should be longer term work.

Reply:
The transformation using delta is the idiomatic way to convert potentially overflowing (and possibly UB) expressions in a conditional operation that avoids overflow. As Coleen states neither form is generally obvious to all readers.

Reply:
I'd prefer to see any change to when/why the ceiling is adjusted deferred to the issue that looks at what it should be adjusted to. Nothing about this ceiling is intuitive or obvious to me.

Reply:
I think it's important that we make sure that the ceiling shouldn't be increased if `monitor_usage` is below the `MonitorUsedDeflationThreshold` in this PR (for the 24 release). But I absolutely agree that we should create a new issue that looks into what it should be adjusted to later.